lib/src/optimism/mod.rs

```diff
@@ -236,122 +284,210 @@ impl<D: BatcherDb> DeriveMachine<D> {
+                        provider_db,
+                        new_op_head_input.clone(),
+                    )
+                    .map(|mut headerless_preflight_data| {
+                        let header = Header {
+                            beneficiary: new_op_head_input.state_input.beneficiary,
+                            gas_limit: new_op_head_input.state_input.gas_limit,
+                            timestamp: new_op_head_input.state_input.timestamp,
+                            extra_data: new_op_head_input.state_input.extra_data.clone(),
+                            mix_hash: new_op_head_input.state_input.mix_hash,
+                            // unnecessary
+                            ..Default::default()
+                        };
+                        headerless_preflight_data.header = Some(header);
+                        headerless_preflight_data
+                    })?;
+
+                    let executable_input: BlockBuildInput<OptimismTxEssence> =
+                        preflight_data.try_into()?;
+                    if let Some(ref mut inputs_vec) = op_block_inputs {
+                        inputs_vec.push(executable_input.clone());
+                    }
+
+                    OptimismStrategy::build_from(&OP_MAINNET_CHAIN_SPEC, executable_input)?
+                        .with_state_hashed()
+                };
+                // guest: ask for receipt about provided block build output (compressed state trie
+                // expected)
+                #[cfg(target_os = "zkvm")]
+                let op_block_output = {
+                    let output = op_block_output_iter.next().unwrap();
+                    // A valid receipt should be provided for block building results
+                    let builder_journal =
+                        to_vec(&output).expect("Failed to encode builder journal");
+                    env::verify(
+                        Digest::from(self.derive_input.block_image_id),
+                        &builder_journal,
+                    )
+                    .expect("Failed to validate block build output");
+                    output
+                };
 
-                    // verify that the new op head mix hash matches the mix hash of the L1 block
-                    {
-                        let l1_epoch_header = self
-                            .derive_input
-                            .db
-                            .get_full_eth_block(op_batch.0.epoch_num)
-                            .context("eth block not found")?
-                            .block_header
-                            .clone();
+                match op_block_output {
+                    BlockBuildOutput::SUCCESS {
+                        hash: new_block_hash,
+                        head: new_block_head,
+                        ..
+                    } => {
+                        // Verify that the built op block matches the payload attributes of the
```

Comment:
Why do we need to do that?
If we cannot trust the Input of the proof, then I guess we'd need to verify everything including the state input etc.

Reply:
it'd have to be part of the success output as well which I think isn't a bad idea